File: R/plotscapeTest.R
#' <Add Title>
#'
#' <Add Description>
#'
#' @import htmlwidgets
#'
#' @export
#' @examples
#' scene(mtcars) |>
#'   add_plot_wrapper("scatter", list(x = "wt", y = "mpg")) |>
#'   add_plot_wrapper("histo", list(x = "disp", y = "_indicator"))
scene <- function(data, width = NULL, height = NULL, elementId = NULL) {
  # forward options using x
  x = list(
    data = data,
    types = c(),
    mappings = list()
  )

  # create widget
  htmlwidgets::createWidget(
    name = 'plotscapeTest',
    x,
    width = width,
    height = height,
    package = 'plotscapeTest',
    elementId = elementId
  )
}

#' @export
add_plot_wrapper <- function(scene, type, mapping) {
  scene$x$types <- c(scene$x$types, type)
  scene$x$mappings <- append(scene$x$mappings, list(mapping))
  scene
}

#' Shiny bindings for plotscapeTest
#'
#' Output and render functions for using plotscapeTest within Shiny
#' applications and interactive Rmd documents.
#'
#' @param outputId output variable to read from
#' @param width,height Must be a valid CSS unit (like \code{'100\%'},
#'   \code{'400px'}, \code{'auto'}) or a number, which will be coerced to a
#'   string and have \code{'px'} appended.
#' @param expr An expression that generates a plotscapeTest
#' @param env The environment in which to evaluate \code{expr}.
#' @param quoted Is \code{expr} a quoted expression (with \code{quote()})? This
#'   is useful if you want to save an expression in a variable.
#'
#' @name plotscapeTest-shiny
#'
#' @export
plotscapeTestOutput <- function(outputId, width = '100%', height = '400px'){
  htmlwidgets::shinyWidgetOutput(outputId, 'plotscapeTest', width, height, package = 'plotscapeTest')
}

#' @rdname plotscapeTest-shiny
#' @export
renderPlotscapeTest <- function(expr, env = parent.frame(), quoted = FALSE) {
  if (!quoted) { expr <- substitute(expr) } # force quoted
  htmlwidgets::shinyRenderWidget(expr, plotscapeTestOutput, env, quoted = TRUE)
}

File: inst/htmlwidgets/plotscapeTest.js
HTMLWidgets.widget({

  name: 'plotscapeTest',

  type: 'output',

  factory: function(el, width, height) {

    // TODO: define shared variables for this instance

    return {

      renderValue: function(x) {

        // TODO: code to render the widget, e.g.
        el.classList.add("graphicDiv")
        el.innerText = PLOTSCAPE.mean(x.data);
        const scene = new PLOTSCAPE.Scene(el, new PLOTSCAPE.DataFrame(x.data))

        x.types.forEach((e, i) => {
          const mapping = x.mappings[i]
          const mappingArray = Object.keys(mapping).map(e => [e, mapping[e]])
          scene.addPlotWrapper(e, new PLOTSCAPE.Mapping(...mappingArray))
        })

      },

      resize: function(width, height) {

        // TODO: code to re-render the widget with a new size

      }

    };
  }
});
